File: src/lib/geo.ts
/**
 * GeoIP enrichment via a locally-bundled MaxMind GeoLite2-City DB
 * (ARCHITECTURE.md §4.6, FR-6, NFR-11, AC-12). Offline, no paid API. Degrades
 * gracefully to {country:null, city:null} if the DB file is absent so the app
 * still runs (the DB is provisioned once for full AC-12 coverage). Used only by
 * the worker.
 */
import { existsSync } from 'node:fs'
import { join } from 'node:path'
import { env } from './env'

export interface GeoResult {
  country: string | null
  city: string | null
}

const EMPTY: GeoResult = { country: null, city: null }

// maxmind's Reader type; loaded lazily.
let reader: { get: (ip: string) => unknown } | null = null
let loadAttempted = false

async function getReader(): Promise<{ get: (ip: string) => unknown } | null> {
  if (loadAttempted) return reader
  loadAttempted = true
  const dbPath = join(process.cwd(), env.geoipDbPath)
  if (!existsSync(dbPath)) {
    console.warn(`[geo] GeoLite2 DB not found at ${dbPath}; geo enrichment disabled.`)
    return null
  }
  try {
    const maxmind = await import('maxmind')
    reader = await maxmind.open(dbPath)
    return reader
  } catch (err) {
    console.warn('[geo] failed to open GeoLite2 DB:', (err as Error).message)
    return null
  }
}

/** Look up country/city for an IP. Never throws. */
export async function geoLookup(ip: string | null | undefined): Promise<GeoResult> {
  if (!ip) return EMPTY
  const r = await getReader()
  if (!r) return EMPTY
  try {
    const rec = r.get(ip) as
      | { country?: { iso_code?: string; names?: { en?: string } }; city?: { names?: { en?: string } } }
      | null
    if (!rec) return EMPTY
    return {
      country: rec.country?.iso_code ?? rec.country?.names?.en ?? null,
      city: rec.city?.names?.en ?? null,
    }
  } catch {
    return EMPTY
  }
}

File: backend/src/services/geo.rs
//! GeoIP enrichment via a locally-bundled MaxMind GeoLite2-City DB — ported from
//! `src/lib/geo.ts`. Offline, no paid API. Degrades gracefully to (None, None)
//! when the DB is absent so the app still runs. Used only by the worker.

use maxminddb::geoip2;

/// Look up (country, city) for an IP. Never panics.
/// `country` prefers the ISO code, then the English country name.
/// `city` is the English city name. Returns (None, None) when:
///   - no DB is loaded,
///   - the IP doesn't parse,
///   - the lookup misses, or
///   - the record has no usable fields.
pub fn lookup(
    db: Option<&maxminddb::Reader<Vec<u8>>>,
    ip: &str,
) -> (Option<String>, Option<String>) {
    let reader = match db {
        Some(r) => r,
        None => return (None, None),
    };
    if ip.is_empty() {
        return (None, None);
    }
    let addr = match ip.parse::<std::net::IpAddr>() {
        Ok(a) => a,
        Err(_) => return (None, None),
    };
    let city: geoip2::City = match reader.lookup(addr) {
        Ok(c) => c,
        Err(_) => return (None, None),
    };

    let country = city
        .country
        .as_ref()
        .and_then(|c| {
            c.iso_code
                .map(|s| s.to_string())
                .or_else(|| c.names.as_ref().and_then(|n| n.get("en").map(|s| s.to_string())))
        });

    let city_name = city
        .city
        .as_ref()
        .and_then(|c| c.names.as_ref())
        .and_then(|n| n.get("en").map(|s| s.to_string()));

    (country, city_name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn none_db_yields_none_none() {
        assert_eq!(lookup(None, "8.8.8.8"), (None, None));
    }

    #[test]
    fn empty_ip_yields_none_none() {
        assert_eq!(lookup(None, ""), (None, None));
    }
}
